Azure Architecture guidance (#6140)
Particular/docs.particular.net: architecture/consistency.md

---
title: Consistency
summary: Achieving consistency in distributed, message-driven systems
reviewed: 2023-07-18
callsToAction: ['solution-architect']
---

## Transactions

Transactions are designed to keep a system in a consistent state at any point in time. This is achieved by ensuring interdependent operations are completed as a single [unit of work](https://en.wikipedia.org/wiki/Unit_of_work). If any single operation fails, then the entire transaction fails and the system is reverted to the state it was in before the transaction started.

### Limitations

There are several challenges and limitations when relying on transactions:

#### Scalability

Scalable systems need to handle high levels of concurrency. Transactions need concurrency control mechanisms to ensure proper isolation between transactions trying to access the same resources. These concurrency control mechanisms can severely impact a resource's performance and introduce bottlenecks that complicate scaling.

> When using horizontal scaling, additional mechanisms are required to ensure the desired level of consistency across all scale units. This typically involves complicated distributed consensus-algorithms. Depending on the consistency guarantees required, many, or even all, nodes of a scaled component need to participate in the transaction, negatively impacting latency and availability.

Author: I couldn't find a good definition from a "trustworthy" source, so how about we link to this MS guidance?

```suggestion
When using [horizontal scaling](https://learn.microsoft.com/en-us/azure/well-architected/scalability/design-scale#understand-horizontal-and-vertical-scaling), additional mechanisms are required to ensure the desired level of consistency across all scale units. This typically involves complicated distributed consensus-algorithms. Depending on the consistency guarantees required, many, or even all, nodes of a scaled component need to participate in the transaction, negatively impacting latency and availability.
```

Note: Implementing message deduplication is risky and error-prone. Small mistakes can lead to unintended behavior or message loss. The [NServiceBus Outbox feature](/nservicebus/outbox/) implements message deduplication and is thoroughly tested and well-documented.

### Side effect checks

Message deduplication assumes a generic deduplication mechanism that works for all incoming messages, regardless of the specific message content. When this is not possible, duplicate message processing can be avoided by checking for expected side effects. These side effects are only known for a specific message type and its related business logic. For example, when `TheFireIsHot` flag is set to true, then there is no need to `TurnOnTheFire`. This approach is less reusable and requires in-depth understanding of the business logic to deal with duplicate messages.

> Additional metadata may be attached to existing side effects to achieve deduplication. For example, a message timestamp (provided by the author of a message) is attached to a database record that is stored while processing that message.

Author: would that be easier to understand?

```suggestion
Additional metadata may be stored to enable side effect checks. For example, a message timestamp or an unique identifier (provided by the author of a message) is attached to a database record that is stored while processing that message.
```

#### Cost

Besides the performance impact of transactions, transactional operations may even be charged extra on certain managed cloud services. Broad usage of transactional consistency may therefore incur significant costs. It is important to understand a cloud vendor's pricing models to estimate the financial cost of using transactions.

#### Timeouts

Transactions impact concurrent operations on a locked resource. This may block other actors and consume additional resources. Therefore transactions should be completed within a short time window to avoid issues. Cloud services typically restrict transaction lifetimes further than traditional on-premises technologies.

#### Scope

Some cloud services like Azure Table Storage, Azure Cosmos DB Table API, and Azure Cosmos DB SQL API offer transactions, but they are usually scoped to a single partition key. When transactions are scoped to a single partition key, atomic storage operations must share the same partition key on all create, read, update, and delete operations.

### Cross-entity transactions

> Databases and queueing technologies may support atomic transactions for a single entity (e.g. a message queue, or a database document) or multiple entities (sometimes also restricted to a single partition). Without cross-entity transactions, queueing systems cannot atomically acknowledge (consume) a message together with sending outgoing messages, which might lead to duplicate messages when retrying a failed message.

Author: this is focused on entities on the same technology, e.g. incoming and outgoing message queue. With cross-entity tx outgoing messages are being rolled back if the incoming message can't be ack'ed. Outbox operates on a different level as it can't roll back and outgoing messages will be dispatched at that point. In that scenario, the Outbox can only prevent the handlers from being invoked again, which on the other hand cross-queue-transactions wouldn't help with.

not against mentioning the outbox here, but I think there are significant differences to take into account so I'm not sure we can state something like "just use the outbox instead"?

> [**Compare message queue transaction capabilities →**](/transports/transactions.md)

Author: our implementation is typically as close to what the transport can support. Is there another place you'd link to?

}}

### Outbox pattern

The [transactional outbox pattern](https://microservices.io/patterns/data/transactional-outbox.html) provides atomically consistent database and message queue operations. Due to the lack of distributed transactions in modern application environments, the outbox pattern is commonly used to satisfy consistency requirements.

The outbox pattern is implemented by storing outgoing messages in the same database as business data. This allows the use of database transaction capabilities to achieve atomic consistency. In the second step, the persisted messages are dispatched to the message queue.

The nature of the outbox pattern may cause outgoing messages to be dispatched multiple times, which requires [idempotency](#idempotency) in receivers.

Note: Implementing the outbox pattern is very risky and error-prone. Small mistakes can lead to unintended behavior and message or data loss. The [NServiceBus outbox feature](/nservicebus/outbox/) implements the outbox pattern, including built-in message deduplication when receiving messages, and is thoroughly tested and well-documented.

## Idempotency

> [Idempotence](https://en.wikipedia.org/wiki/Idempotence) ensures that processing a message multiple times has the same effect as processing it only once. This means that processing a duplicate message shouldn’t cause any unintended side effects.

Author: added a link 👍

When using [horizontal scaling](https://learn.microsoft.com/en-us/azure/well-architected/scalability/design-scale#understand-horizontal-and-vertical-scaling), additional mechanisms are required to ensure the desired level of consistency across all scale units. This typically involves complicated distributed consensus algorithms. Depending on the consistency guarantees required, many, or even all, nodes of a scaled component need to participate in the transaction, negatively impacting latency and availability.

#### Cost

Besides the performance impact of transactions, transactional operations may even be charged extra on certain managed cloud services. Broad usage of transactional consistency may therefore incur significant costs. It is important to understand a cloud vendor's pricing models to estimate the financial cost of using transactions.

#### Timeouts

> Transactions impact concurrent operations on a locked resource. This may block other actors and consume additional resources. Therefore, transactions should be completed within a short time window to avoid issues. Cloud services typically restrict shorter transaction lifetimes than traditional on-premises technologies.

Reviewer: ```suggestion
Transactions impact concurrent operations on a locked resource. This may block other actors and consume additional resources. Therefore, transactions should be completed within a short time window to avoid issues. Cloud services typically restrict transaction lifetimes more than traditional on-premises technologies.
```

The outbox pattern is implemented by storing outgoing messages into the same database as the business operations. This allows the use of database native transaction capabilities to achieve atomic consistency. In a second step, the persisted messages are dispatched to tbe message queue.

The nature of the outbox pattern might cause outgoing messages to be dispatched multiple times. This should be taken into consideration by leveraging [idempotency mechanisms](#idempotency) on the receiver.

Note: Implementing the outbox pattern is very risky and error-prone. Small mistakes can lead to unintended behavior and message or data loss. Consider the the [NServiceBus Outbox feature](/nservicebus/outbox/) which implements the Outbox pattern, including built-in message deduplication when receiving messages. The NServiceBus Outbox is used in production by hundreds of customers, thoroughly tested and well documented.

> #### Transactional Session
> 
> The TransactionalSession is an NServiceBus feature specifically designed to bring Outbox functionality to system boundaries like web servers that require outgoing message operations and database operations to be consistent.

Author: ```suggestion
```

Thinking more about this, I feel we can remove this since the transactional session is also just an implementation of the outbox pattern. The outbox documentation will point to the tx session too.

> ### Session Management
> 
> When NServiceBus is configured with a persistent storage, NServiceBus will manage the lifecycle of (transactional) storage sessions. Application code can access the managed session to work with business data:
> 
> ```mermaid
> sequenceDiagram
>     Queue->>NSB: ReceiveMessage
>     NSB->>Storage: Open session
>     NSB->>Application: Invoke message handler
>     Application->>Storage: Store business data
>     Application-->>NSB: message handler finished
>     NSB->>Storage: Store workflow/outbox data
>     NSB->>Storage: Commit transaction
>     NSB-xStorage: Close session
>     NSB-->>Queue: Ack message
> ```

Author: ```suggestion
```

I think this might not be valuable enough from an architectural perspective and is too NSB specific for this section.

Reviewer: Is this anywhere else?

Author: no but we could consider adding this to the saga documentation instead.

Reviewer: As discussed, we'll remove this from this PR and raise an issue instead.
